### PR TITLE
VLC-X11: new port, CLI version of vlc 3.0.20

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

Latest VLC3 headless.

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.6
Xcode 3.2

macOS 14.4.1
Xcode 15.3

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: multimedia/VLC-X11/files/0005-configure-do-not-err-out-on-implicit-function-decl.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
+From 8d870f73f214f220bf7415628f8f5a8fd8b012e2 Mon Sep 17 00:00:00 2001
+From: Sergey Fedorov <vital.had@gmail.com>
+Date: Mon, 15 Apr 2024 20:32:15 +0800
+Subject: [PATCH] configure: do not err out on implicit function decl
```
Comment (ryandesign) [inline multimedia/VLC-X11/files/0005-configure-do-not-err-out-on-implicit-function-decl.patch:4]: Disabling the implicit function declaration error is not recommended. Allowing implicit function declarations may lead to incorrect program execution on Apple Silicon processors.

File: multimedia/VLC-X11/Portfile [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+use_autoreconf          yes
+
+# The purpose of this port is to be as compatible as possible and work on every macOS
+# version, including PowerPC systems. Priority is given to legacy systems,
+# since on the current ones VLC3 can be built with GUI and all modules.
+# We do not want a dozen versions here with complex conditional fallback code.
+# Some choices below may not be the first best for newer macOS versions.
+# If this does not fit your aims, please rather install a standard VLC3
+# either from Macports or the developer.
```
Comment (ryandesign) [inline multimedia/VLC-X11/Portfile:38]:
```suggestion
# either from MacPorts or the developer.
```

File: multimedia/VLC-X11/Portfile [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+# already in VLC 2.2.8. However, it may be possible to fix both.
+
+# Macros are modified to be actually useful: HAVE_OSX becomes a configurable option
+# instead of a coerced one; HAVE_DARWIN is used as a macOS baseline.
+# That is, everything macOS-specific which can be supported from 10.5+
+# should go under HAVE_DARWIN. Anything which requires modern Xcode
+# or newer SDK and does not have a fallback implemented should go inside HAVE_OSX,
+# and we conveniently disable all that breaking stuff via --disable-osx arg.
+# Notice, this is not the same as --disable-macosx* of the upstream;
+# despite a suggestive wording, the latter fails to actually disable OSX code.
```
Comment (ryandesign) [inline multimedia/VLC-X11/Portfile:52]:
```suggestion
# despite a suggestive wording, the latter fails to actually disable macOS code.
```

File: multimedia/VLC-X11/Portfile [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+                        0013-Do-not-build-a-broken-libbonjour_plugin.patch \
+                        0014-Fix-environ-define.patch \
+                        0015-libauhal-fix-the-build.patch \
+                        0016-libupnp-fix-condition-for-frameworks-linking.patch
+
+if {[string match *gcc* ${configure.compiler}]} {
+    patchfiles-append   0017-Trash-some-breaking-flags.patch
+}
+
+# FIXME: this patch is a fallback to an older code.
```
Comment (ryandesign) [inline multimedia/VLC-X11/Portfile:74]:
```suggestion
# FIXME: this patch is a fallback to older code.
```

File: multimedia/VLC-X11/Portfile [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+# VLC has a zoo of options, and they change over time.
+# If something is found to be redundant or, on the contrary,
+# lacking, please help to improve the port.
+# It is desirable not to make any potentially breaking changes
+# here without proper testing on systems of concern,
+# which include PowerPC and legacy Intel.
+
+# TODO: we may enable libplacebo once a legacy fallback for it
+# gets merged in Macports. libnotify and samba could be supported
```
Comment (ryandesign) [inline multimedia/VLC-X11/Portfile:206]:
```suggestion
# gets merged in MacPorts. libnotify and samba could be supported
```

**Review by ryandesign — COMMENTED** [review]
So many patches. ☹️ What is the status of getting those patches accepted into VLC upstream? We do not want to be the maintainers of large patchsets.

> Latest VLC3 headless.

What does "headless" mean here? The usage of the term "headless" that I'm familiar with is without a display, but how could a video playback program have no display?

**Comment by barracuda156** [comment]
@ryandesign 

1. I made changes mostly per-patch, so it is a bit more readable. These can be squashed, like it is done in some other ports. (And there are ports with large patches.)
2. I making it my port, and I can maintain it.
3. Upstream does not care about supporting older macOS for many years, and VLC was broken on those since 2.0.x. I won’t waste time trying to convince anyone there, to be honest.
4. The VLC itself has no GUI here, it should be started and operated via CLI. Video works fine via X window though.
You can try `vlc file://${file_on_the_disk}` – that works for me.

File: multimedia/VLC-X11/files/0005-configure-do-not-err-out-on-implicit-function-decl.patch (hunk 1 above)
Comment (barracuda156) [reply multimedia/VLC-X11/files/0005-configure-do-not-err-out-on-implicit-function-decl.patch:4]: Ok, I will restrict the patch to older systems and verify it builds on Sonoma without it.

File: multimedia/VLC-X11/Portfile (hunk 3 above)
Comment (barracuda156) [reply multimedia/VLC-X11/Portfile]: Commit title is wrong, but I will squash these anyway. (Just woke up and reviewing this from mobile.)

**Comment by barracuda156** [comment]
@ryandesign I will squash a few commits which change macros, that gonna reduce total number by a few without affecting readability.

**Comment by barracuda156** [comment]
@ryandesign I have squashed and sorted out patches, I think it should be good to merge.

I have verified that the build works and video with sound works fine on Sonoma and my SL PowerPC.

**Comment by barracuda156** [comment]
@ryandesign Could we perhaps merge this?

**Comment by barracuda156** [comment]
@ryandesign @herbygillot @pmetzger I don’t think there are reasons to hold this now.
We are not getting Qt5 on older systems any time soon. But this way we at least get the current version of VLC functioning there.

**Comment by cellularmitosis** [comment]
wow this is great @barracuda156 !

**Comment by barracuda156** [comment]
Switched it to `ffmpeg6` now.

**Comment by barracuda156** [comment]
@cellularmitosis First of all, I have to apologize for not getting this into MacPorts, but I have no control over what gets merged.

Then, I really advise to abandon vlc, mplayer and mpv on PowerPC systems and use `QMPlay2` port (it is available in MacPorts master). QMPlay2 does not rely on X server and SDL and instead uses Qt, which is faster and far more reliable.

It takes too much effort to bring VLC to a functional state, fixing all bugs which upstream keeps adding in LOL
Advantages are questionable at best: yes, it has more plugins like rendering video into text art, but the whole thing is not reliable and not too convenient without GUI.

**Pull request closed without merging**